### PR TITLE
[BACKLOG-37078] Keyboard Navigation and Focus Order for Browse Files (Folder Dialogs)

### PR DESCRIPTION
**Pull request by jsaraf**

### CODE REVIEW

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 10m 42s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl user-console
```
---
:ok_hand: All tests passed!
 
Tests run: **12**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/platform/job/PR-5213/1/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by jsaraf** [comment]
@pentaho/wcag

File: user-console/src/main/java/org/pentaho/mantle/client/commands/DeleteFolderCommand.java [hunk 1]
```diff
@@ -133,14 +137,31 @@ public void onResponseReceived( Request request, Response response ) {
...
+                          if(isActiveElement(this.okButton.getElement())){
+                            event.preventDefault();
+                            this.hide();
+                            break;
+                          }
+                        }
+                      }
+                    }
+                  }
+                };
```
Comment (dcleao) [inline user-console/src/main/java/org/pentaho/mantle/client/commands/DeleteFolderCommand.java:164]: Recently the ENTER preview handling was removed. Should not be needed anymore.

File: user-console/src/main/java/org/pentaho/mantle/client/commands/DeleteFolderCommand.java [hunk 2]
```diff
@@ -151,7 +172,9 @@ public void onResponseReceived( Request request, Response response ) {
     folderDeleteWarningDialogBox.center();
 
   }
-
+  public static native boolean isActiveElement( Element e )/*-{
+    return $doc.activeElement === e;
+ }-*/;
```
Comment (dcleao) [inline user-console/src/main/java/org/pentaho/mantle/client/commands/DeleteFolderCommand.java:177]: Not needed anymore, but, btw, already exists as `ElementUtils.isActiveElement(.)`.

File: user-console/src/main/java/org/pentaho/mantle/client/commands/ImportFileCommand.java [hunk 3]
```diff
@@ -87,6 +89,10 @@ public void okPressed() {
     };
     importDialog.setCallback( callback );
     importDialog.center();
+    Focusable focusable = ElementUtils.findFirstKeyboardFocusableDescendant( importDialog.getForm().getWidget() );
+    if( focusable != null ){
+      focusable.setFocus( true );
+    }
```
Comment (dcleao) [inline user-console/src/main/java/org/pentaho/mantle/client/commands/ImportFileCommand.java:95]: The default dialog code should already place the focus on the "browse" button.
A bug that cause this to fail when the first element was disabled (your case) has been fixed already.
Please, rebase on `wcag` latest and retest.

File: user-console/src/main/java/org/pentaho/mantle/client/commands/NewFolderCommand.java [hunk 4]
```diff
@@ -100,7 +100,7 @@ protected void performOperation( boolean feedback ) {
     event.setAction( this.getClass().getName() );
 
     final TextBox folderNameTextBox = new TextBox();
-    folderNameTextBox.setTabIndex( 1 );
+    folderNameTextBox.setTabIndex( 0 );
```
Comment (dcleao) [inline user-console/src/main/java/org/pentaho/mantle/client/commands/NewFolderCommand.java:103]: AFAIK, tab index 0 is the default for text boxes. So you should just remove the statement.

File: user-console/src/main/java/org/pentaho/mantle/client/solutionbrowser/fileproperties/FilePropertiesDialog.java [hunk 5]
```diff
@@ -216,6 +218,14 @@ public void showTab( Tabs tab ) {
...
     }
   }
+  @Override
+  public void center() {
+    super.center();
+    Focusable focusable = ElementUtils.findFirstKeyboardFocusableDescendant( this.propertyTabs );
+    if ( focusable != null ) {
+      focusable.setFocus( true );
+    }
+  }
```
Comment (dcleao) [inline user-console/src/main/java/org/pentaho/mantle/client/solutionbrowser/fileproperties/FilePropertiesDialog.java:228]: Can you explain why this is needed? Why does the existing generic code fail?
Also, _if_ you'd really need to do this, ideally, you'd do it before super.center and by passing the intended widget to `setFocusWidget(.)`.

**Review by dcleao — CHANGES_REQUESTED** [review]

File: user-console/src/main/java/org/pentaho/mantle/client/commands/ImportFileCommand.java (hunk 3 above)
Comment (jsaraf) [reply user-console/src/main/java/org/pentaho/mantle/client/commands/ImportFileCommand.java:95]: Upon pulling in latest code and testing, found that without this piece of code, focus is placed on the "cancel" button replicating the behaviour of the bug.

**Comment by buildguy** [comment]
### :x: Build failed in 8h 9m 5s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl user-console
```
---
 :exclamation: No tests found!
 ---
### Errors:
<details><summary><b>Filtered log</b> (click to expand)</summary>
<p>

```
script returned exit code 1
```
 
</p>
</details>
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 11m 19s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl user-console
```
---
:ok_hand: All tests passed!
 
Tests run: **12**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/platform/job/PR-5213/3/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

File: user-console/src/main/java/org/pentaho/mantle/client/commands/ImportFileCommand.java (hunk 3 above)
Comment (dcleao) [reply user-console/src/main/java/org/pentaho/mantle/client/commands/ImportFileCommand.java:95]: Please see https://github.com/pentaho/pentaho-commons-gwt-modules/pull/843 which addresses this bug.

File: user-console/src/main/java/org/pentaho/mantle/client/solutionbrowser/fileproperties/FilePropertiesDialog.java (hunk 5 above)
Comment (dcleao) [reply user-console/src/main/java/org/pentaho/mantle/client/solutionbrowser/fileproperties/FilePropertiesDialog.java:228]: Please see https://github.com/pentaho/pentaho-commons-gwt-modules/pull/843 which addresses this bug.

File: user-console/src/main/java/org/pentaho/mantle/client/commands/ImportFileCommand.java (hunk 3 above)
Comment (dcleao) [reply user-console/src/main/java/org/pentaho/mantle/client/commands/ImportFileCommand.java:95]: Please see https://github.com/pentaho/pentaho-commons-gwt-modules/pull/843 which addresses this bug.

**Comment by dcleao** [comment]
@jsaraf  Please **rebase** your changes on top of _wcag_ instead of **merging**, like you did above with the commit https://github.com/pentaho/pentaho-platform/pull/5213/commits/4b3a759181fb46b4a5bf08ab7023f094a6e6af17.

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 11m 5s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl user-console
```
---
:ok_hand: All tests passed!
 
Tests run: **12**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/platform/job/PR-5213/4/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 10m 44s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl user-console
```
---
:ok_hand: All tests passed!
 
Tests run: **12**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/platform/job/PR-5213/5/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

File: user-console/src/main/java/org/pentaho/mantle/client/solutionbrowser/fileproperties/FilePropertiesDialog.java (hunk 5 above)
Comment (jsaraf) [reply user-console/src/main/java/org/pentaho/mantle/client/solutionbrowser/fileproperties/FilePropertiesDialog.java:228]: I tried suggested change but it would break the behaviour so I have kept it as is until further instruction.

**Comment by jsaraf** [comment]
@dcleao Please check and suggest any changes required.

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 11m 31s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl user-console
```
---
:ok_hand: All tests passed!
 
Tests run: **12**, Failures: **0**, Skipped: **0** &nbsp;&nbsp; [Test Results](https://wingman.orl.eng.hitachivantara.com/job/projects/job/platform/job/PR-5213/6/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

File: user-console/src/main/java/org/pentaho/mantle/client/commands/DeleteFolderCommand.java [hunk 6]
```diff
@@ -133,6 +134,7 @@ public void onResponseReceived( Request request, Response response ) {
                 dialogBox.center();
                 event.setMessage( Messages.getString( "couldNotDeleteFolder" ) );
                 EventBusUtil.EVENT_BUS.fireEvent( event );
+                dialogBox.hide();
```
Comment (dcleao) [inline user-console/src/main/java/org/pentaho/mantle/client/commands/DeleteFolderCommand.java:137]: Hiding the dialogBox immediately after showing it (using `center()`)?
Why is this necessary?

File: user-console/src/main/java/org/pentaho/mantle/client/commands/ImportFileCommand.java [hunk 7]
```diff
@@ -20,6 +20,7 @@
 
 package org.pentaho.mantle.client.commands;
 
+
```
Comment (dcleao) [inline user-console/src/main/java/org/pentaho/mantle/client/commands/ImportFileCommand.java:23]: No changes in file.

**Review by dcleao — CHANGES_REQUESTED** [review]

File: user-console/src/main/java/org/pentaho/mantle/client/commands/DeleteFolderCommand.java (hunk 6 above)
Comment (jsaraf) [reply user-console/src/main/java/org/pentaho/mantle/client/commands/DeleteFolderCommand.java:137]: Removing the line causes the dialog to remain on the screen even after pressing the "OK" button.

Comment (dcleao) [reply user-console/src/main/java/org/pentaho/mantle/client/commands/DeleteFolderCommand.java]: But this way, the dialog would not even be shown...

Comment (jsaraf) [reply user-console/src/main/java/org/pentaho/mantle/client/commands/DeleteFolderCommand.java]: After rebasing I have observed that the focus is set to `No` instead of `Yes, Move to Trash` upon opening the dialog, also the error dialog is now taking two keyboard/mouse inputs to close.
.

Comment (jsaraf) [reply user-console/src/main/java/org/pentaho/mantle/client/commands/DeleteFolderCommand.java]: @dcleao can you look into this? Thank you.

Comment (dcleao) [reply user-console/src/main/java/org/pentaho/mantle/client/commands/DeleteFolderCommand.java]: The "No" button being selected by default is the new default, as it is the safest option.

It's arguable that, for this dialog, the default button could be "Yes, Move to Trash", given moving to trash is a reversible action.

If desired, the default can be changed, by using the dialog's `setFocusWidget( . )` method. However, the `okButton` field is protected, and as such, it is currently necessary to create a subclass (it can be an anonymous one).

Comment (dcleao) [reply user-console/src/main/java/org/pentaho/mantle/client/commands/DeleteFolderCommand.java]: I'm not sure what you mean by "error dialog"? This dialog is a confirmation dialog. If you are referring to it, when the default is "No", accepting the action will require two keyboard inputs: a Tab and an Enter. This is expected, if we agree that the default button should be "No".
